go/services/connector/main.go

```diff
@@ -0,0 +1,620 @@
+		if _, exists := deviceData.trackedAssets[assetName]; !exists {
+			deletionChan := make(AssetDeletionChan)
+			deviceData.trackedAssets[assetName] = deletionChan
+
+			f.logger.Info("New asset detected",
+				"device", deviceRef.String(),
+				"asset", assetName)
+
+			deviceData.assetCreationTx <- AssetNotification{
+				Asset:         assetRef,
+				DeletionToken: deletionChan,
+			}
+		}
+	}
+}
+
+// NewDeviceEndpointCreateObservation creates a new DeviceEndpointCreateObservation.
+func NewDeviceEndpointCreateObservation(
+	ctx context.Context,
+	pollInterval time.Duration,
+	logger *slog.Logger,
+) (*DeviceEndpointCreateObservation, error) {
+	mountPath, err := GetMountPath()
+	if err != nil {
+		return nil, err
+	}
+
+	fileMountMap := NewFileMountMap(mountPath, logger)
+	connMonitor := NewConnectionMonitor()
+	observeCtx, cancel := context.WithCancel(ctx)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to create watcher: %w", err)
+	}
+
+	// Always watch the parent directory so that we can detect the mount point being created/removed.
+	parentDir := filepath.Dir(mountPath)
+	if err := watcher.Add(parentDir); err != nil {
+		watcher.Close()
+		cancel()
+		return nil, fmt.Errorf("failed to watch parent directory: %w", err)
+	}
+
+	// If the mount currently exists, start watching it.
+	if _, err := os.Stat(mountPath); err == nil {
+		if err := watcher.Add(mountPath); err != nil {
+			watcher.Close()
+			cancel()
+			return nil, fmt.Errorf("failed to watch mount path: %w", err)
+		}
+	}
+
+	observation := &DeviceEndpointCreateObservation{
+		mountPath:    mountPath,
+		deviceChan:   fileMountMap.createDeviceTx,
+		fileMountMap: fileMountMap,
+		ctx:          observeCtx,
+		cancel:       cancel,
+		pollInterval: pollInterval,
+		connMonitor:  connMonitor,
+		logger:       logger,
+		watcher:      watcher,
+	}
+
+	// Initialize with existing devices
+	_, err = os.Stat(mountPath)
+	switch {
+	case err == nil:
+		// Mount path exists, check for devices
+		devices, err := getDeviceEndpointNames(mountPath)
+		if err != nil {
+			cancel()
+			watcher.Close()
+			return nil, err
+		}
+
+		fileMountMap.UpdateDeviceEndpoints(devices)
+		connMonitor.SetConnected()
+	case os.IsNotExist(err):
+		// Mount path doesn't exist yet
+		connMonitor.SetDisconnected()
+	default:
+		cancel()
+		watcher.Close()
+		return nil, &Error{Message: "Failed to access mount path", Cause: err}
+	}
+
+	// Start watcher loop
+	go observation.watchLoop()
+
+	return observation, nil
+}
+
+// watchLoop processes fsnotify events and updates the file mount map.
+func (o *DeviceEndpointCreateObservation) watchLoop() {
+	defer o.watcher.Close()
+
+	for {
+		select {
+		case <-o.ctx.Done():
+			return
+		case event, ok := <-o.watcher.Events:
+			if !ok {
+				return
+			}
+
+			// Handle creation or removal of the mount path itself.
+			if event.Name == o.mountPath && (event.Op&fsnotify.Create) != 0 {
```

Comment:
Huh. TIL this was a bitfield. We might want to review the SAT logic to make sure it's still valid in that light.

That being said, it's probably cleaner to use `event.Has` instead of mucking with the bits directly.